### PR TITLE
feat: adjust axis limits when using reference lines outside of the chart

### PR DESCRIPTION
**Pull request by rephus**
Closes: #4117

### Description:

Things in scope: 

- [x]  We can adjust axis on bottom (positive and negative) 
image

- [x] We can adjust axis on the left (postivie and negative) 

image

- [x] We can adjust a single axis on the right 

image

- [x] If there is a custom min/max value, we don't adjust the axis 

image

Things out of scope:

- [ ] Adjust axis when they are flipped. We need to fix this first: https://github.com/lightdash/lightdash/issues/4141
- [ ] Top axis . Same as before. 
- [ ] Having multiple axes (left and right) . it should be automatically fixed after this https://github.com/lightdash/lightdash/issues/4143
It kind of works despite of the bug, but it is not reliable 
image

### CODE REVIEW

File: packages/frontend/src/hooks/echarts/useEcharts.ts [hunk 1]
```diff
@@ -640,12 +738,14 @@ const getEchartAxis = ({
                       })
                     : xAxisConfiguration?.[0]?.name ||
                       (xAxisItem ? getItemLabel(xAxisItem) : undefined),
-                min: validCartesianConfig.layout.flipAxes
+                min: !validCartesianConfig.layout.flipAxes
```
Comment (rephus) [inline packages/frontend/src/hooks/echarts/useEcharts.ts:741]: I think this was a typo, on the bottom axis we should apply the values if it is not flipped

File: packages/frontend/src/hooks/echarts/useEcharts.ts [hunk 2]
```diff
@@ -252,6 +253,88 @@ const removeEmptyProperties = <T = Record<any, any>>(obj: T | undefined) => {
...
+        axis: string,
+        fieldId: string | undefined,
+    ): number[] => {
+        const values = series.flatMap((serie) => {
+            const serieFieldId =
+                axis === 'yAxis' ? serie.encode.yRef : serie.encode.xRef;
+            if (serieFieldId.field !== fieldId) return [];
+
+            if (!serie.markLine) return [];
+            return serie.markLine?.data.reduce<number[]>((acc, data) => {
```
Comment (rephus) [inline packages/frontend/src/hooks/echarts/useEcharts.ts:292]: this reduce should make it compatible with multiple reference lines (separate branch)

**Comment by ZeRego** [comment]
Still not seeing the reference line when axis are flipped 

Screenshot 2023-01-11 at 11 25 32

**Comment by rephus** [comment]
> Still not seeing the reference line when axis are flipped

Yep, flipping doesn't work yet, I added that as out of scope on the description of this PR 

> Reference lines do not work when flipping axis  #4141

**Comment by ZeRego** [comment]
> > Still not seeing the reference line when axis are flipped
> 
> Yep, flipping doesn't work yet, I added that as out of scope on the description of this PR
> 
> > Reference lines do not work when flipping axis  #4141

Sorry my bad

File: packages/frontend/src/hooks/echarts/useEcharts.ts (hunk 1 above)
Comment (ZeRego) [reply packages/frontend/src/hooks/echarts/useEcharts.ts:741]: min/max seems broken in your PR.

demo:
Screenshot 2023-01-11 at 14 38 10

PR env:
Screenshot 2023-01-11 at 14 39 09

Comment (rephus) [reply packages/frontend/src/hooks/echarts/useEcharts.ts]: Yep, you're right, it was not a typo. It was there to apply the limit when flipped. I fixed it

**Comment by TuringLovesDeathMetal** [comment]
This looks great @rephus !!

Only tiny thing I noticed is that if you have the reference line below min value of the chart on the x-axis, then the axis ticks shift onto the chart (instead of staying on the axis). Let me know if that doesn't make sense 😅 (also not sure if this is something we can change, or if this is an echarts thing)

Screenshot 2023-01-13 at 10 02 14

**Comment by rephus** [comment]
> Only tiny thing I noticed is that if you have the reference line below min value of the chart on the x-axis, then the y-axis ticks shift onto the chart (instead of staying on the axis). Let me know if that doesn't make sense sweat_smile (also not sure if this is something we can change, or if this is an echarts thing)

I think that is an echarts thing, we can also get this if we manually increase the limit (without reference line) . The tics stay on 0 

image

**Review by ZeRego — APPROVED** [review]

**Pull request merged**